[FLINK-31481][table] Support enhanced show databases syntax

jeyhunkarimov opened the pull request:
## What is the purpose of the change

Support enhanced show functions syntax described in [FLIP-297](https://cwiki.apache.org/confluence/display/FLINK/FLIP-297%3A+Improve+Auxiliary+Sql+Statements)


## Brief change log

Changelog from:
```
SHOW DATABASES
```

to:

```
SHOW DATABASES [ ( FROM | IN ) catalog_name] [ [NOT] (LIKE | ILIKE) <sql_like_pattern> ]
```



## Verifying this change

  - Extended integration tests `flink-table/flink-sql-client/src/test/resources/sql/catalog_database.q`
  - Extended unit test `flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java -> testShowDataBases()`
  - Added unit tests in `flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/operations/SqlOtherOperationConverterTest.java -> testShowDatabases()`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency):  no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

jeyhunkarimov commented:
Hi @luoyuxia @chucheng92 could you please review the PR? Thanks!

jeyhunkarimov commented:
Hi @MartijnVisser could you please check this PR?

jeyhunkarimov commented:
Hi @jnh5y @dawidwys , could you guys check this PR?

jnh5y commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowDatabasesOperation.java:
```diff
@@ -20,26 +20,108 @@
+                                        if (likeType == LikeType.ILIKE) {
+                                            return notLike
+                                                    != SqlLikeUtils.ilike(row, likePattern, "\\");
+                                        } else {
```
I'd be tempted to suggest checking for the other type here.

jeyhunkarimov replied:
done

jnh5y commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowDatabasesOperation.java:
```diff
@@ -20,26 +20,108 @@
+                ctx.getCatalogManager().getCatalogOrThrowException(cName).listDatabases().stream()
                         .sorted()
                         .toArray(String[]::new);
+
+        if (likeType != null) {
+            databases =
+                    Arrays.stream(databases)
```
You could hold on to the `ctx.getCatalogManager().getCatalogOrThrowException(cName).listDatabases().stream()` as a stream and only sort it once after you've (potentially) filtered it.

jeyhunkarimov replied:
done

jnh5y commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/converters/SqlShowDatabasesConverter.java:
```diff
@@ -0,0 +1,58 @@
+        } else {
+            CatalogManager catalogManager = context.getCatalogManager();
+            String[] fullCatalogName = sqlShowDatabases.getCatalog();
+            if (fullCatalogName.length > 1) {
```
Two things:

1.  I wonder if this check should be done sooner.
2. Does the client testing allow for dealing with errors?  If so, adding this case to the `catalog_database.q` would be nice.

jeyhunkarimov replied:
Yes, the `length` check is moved to the constructor of `SqlShowDatabases`, and I added a test case to catch the exception in `catalog_database.q `

jnh5y replied:
Cool.  @dawidwys does that make sense?  Or should we avoid having constructors throw exceptions?  (I'm not sure what the best practice is here.)

jnh5y commented on flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowDatabases.java:
```diff
@@ -29,14 +31,55 @@
+            boolean notLike) {
         super(pos);
+        this.preposition = preposition;
+        this.catalogName =
```
For example, I think it may make sense to check the `SqlIdentifier`  `catalogName` here to see if it is has more than one part.

jeyhunkarimov replied:
Yes, good catch

chucheng92 commented:
> This is looking pretty good. Since implementing FLIP-297 is assigned to @chucheng92, we should get his input as well!
> 
> @jeyhunkarimov have you signed up for the Flink JIRA? You may need to request access on the Flink dev list. (I tried to find instructions, but I am not seeing them immediately.)

@jeyhunkarimov good work! Currently i'm busy with other work, feel free to assign to yourself.

jeyhunkarimov commented:
Thanks for the comment @chucheng92 and thank you for the review @jnh5y .

I addressed your comments. Could you please check if you have any other comments?

P.s I have jira account (jID: jeyhunkarimov). I can assign tasks to myself in other apache projects (e.g., kafka). However, in Flink this is not possible. I will ask for the required permissions via dev mailing list.

jeyhunkarimov commented:
Hi @jnh5y could you please check if your comments are addressed? Thanks!

jnh5y commented on flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowDatabases.java:
```diff
@@ -29,14 +33,64 @@
+        if (this.catalogName != null && this.catalogName.names.size() > 1) {
+            throw new ParseException(
+                    String.format(
+                            "Show databases from/in identifier [ %s ] format error, catalog cannot contain dot character.",
```
As something that I'm unsure about... I wonder if a catalog name can contain a `.` if it quoted with backticks.

If so, then this statement is inaccurate.

dawidwys replied:
I believe you can quote any character.

dawidwys replied:
Given the responses, could you please update the exception to e.g.

```
"Show databases from/in identifier [ %s ] format error, catalog must be a single part identifier."
```

jnh5y commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowDatabasesOperation.java:
```diff
@@ -20,26 +20,108 @@
+        }
+
+        return buildStringArrayResult(
+                "database name", databases.stream().sorted().toArray(String[]::new));
```
Thanks for improving this!  I'll leave it up to @dawidwys if he has any other ideas here.  

I'm tempted to suggest something that doesn't collect the intermediate list.  (On one hand, this code isn't in an inner loop, so optimization may not be strictly necessary.  On another hand, it is useful to think through minimizing intermediate work!)

dawidwys replied:
If `databases` is a `Stream<String>` you wouldn't need to collect the filtering results.

```
Stream<String> databases =
                ctx.getCatalogManager().getCatalogOrThrowException(cName).listDatabases().stream();

        if (likeType != null) {
            databases =
                    databases
                            .filter(
                                    row -> {
                                        if (likeType == LikeType.ILIKE) {
                                            return notLike
                                                    != SqlLikeUtils.ilike(row, likePattern, "\\");
                                        } else if (likeType == LikeType.LIKE) {
                                            return notLike
                                                    != SqlLikeUtils.like(row, likePattern, "\\");
                                        }
                                        return false;
                                    });
        }
```

jnh5y commented:
> Hi @jnh5y could you please check if your comments are addressed? Thanks!

Hi @jeyhunkarimov, sorry to be slow to look again.  Things are looking good.  @dawidwys would you be willing to take a quick read through.  I've left some comments where I think you could provide some insight about what needs to change for this work to be merged.

dawidwys commented on docs/content.zh/docs/dev/table/sql/show.md:
````diff
@@ -507,10 +507,21 @@ SHOW CURRENT CATALOG
 
 ```sql
-SHOW DATABASES
+SHOW DATABASES [ ( FROM | IN ) catalog_name] [ [NOT] (LIKE | ILIKE) <sql_like_pattern> ]
````
Out of curiosity. Do you know chinese?

jeyhunkarimov replied:
Not really :) Used some translating apps + checked existing chinese documentation, so that this feature is documented somehow.

dawidwys replied:
Let's keep it, but I'd prefer not to do that. I don't speak chinese and I don't feel comfortable committing something I don't understand. There is no way for me to say if you did not write something offending there. I believe you did not, but I can not tell for sure.

jeyhunkarimov replied:
Agree. Removed

dawidwys commented on flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlShowDatabases.java:
```diff
@@ -46,11 +100,40 @@ public SqlOperator getOperator() {
     public void unparse(SqlWriter writer, int leftPrec, int rightPrec) {
-        writer.keyword("SHOW DATABASES");
+        String resKeyword = "SHOW DATABASES";
+        writer.keyword(resKeyword);
```
nit: inline `resKeyword`?

dawidwys commented on flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/ShowDatabasesOperation.java:
```diff
@@ -20,26 +20,108 @@
 @Internal
 public class ShowDatabasesOperation implements ShowOperation {
 
+    private final String preposition;
```
Is there a difference whether it is `FROM` or `IN` ? If there is no difference, could we parse both, but use one of the two when unparsing? We wouldn't need to store the original preposition.

jeyhunkarimov replied:
I see what you mean. IMO we need to keep `preposition` for 

- the correct summary string (`ShowDatabasesOperation::asSummaryString`)
- therefore, unparsing accordingly

dawidwys replied:
I don't get the argumentation. Could you please elaborate? `asSummaryString` prints a summary of the operation, I don't see why we need the exact same preposition that was used. I could even see the summary as `("ShowDatabases(catalog='abc', like='...', ....)`.

Even more I don't see a problem with unparsing to a different preposition as long as the effect is the same. In the end it makes no difference at all which one was used.

jeyhunkarimov replied:
If we are flexible with the summary string, then I completely agree

dawidwys commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/operations/converters/SqlShowDatabasesConverter.java:
```diff
@@ -0,0 +1,51 @@
+                    sqlShowDatabases.getLikeSqlPattern(),
+                    sqlShowDatabases.isNotLike());
+        } else {
+            CatalogManager catalogManager = context.getCatalogManager();
+            String[] fullCatalogName = sqlShowDatabases.getCatalog();
+            String catalogName =
+                    fullCatalogName.length == 0
+                            ? catalogManager.getCurrentCatalog()
+                            : fullCatalogName[0];
```
Do we need to get the current catalog at this point? You handle that during the execution already, don't you?

jeyhunkarimov replied:
Good point. 
As you pointed out, we already  get the `catalog` during the execution (`ShowDatabasesOperation::execute(Context ctx)`). We get the catalog from the `ctx` object.

However, we also use `catalogName` inside `ShowDatabasesOperation::asSummaryString()` that takes no arguments (i.e., no access to the `Context`). As a result,`asSummaryString` method will have no catalog name in its result.

dawidwys replied:
Won't it be more correct to not print the catalog in the summary string if it was not given by the user?

jeyhunkarimov replied:
That is also what we do. If no catalog is provided, summary string outputs no catalog name:

```
        final String sql1 = "SHOW DATABASES";
        assertShowDatabases(sql1, sql1);
```

The thing is if a user provides a catalog name, it does not necessarily execute the `ShowDatabaseOperation::execute(Context ctx)` method. For example, please see the below code snipped I used in tests. 

```
    private void assertShowDatabases(String sql, String expectedSummary) {
        Operation operation = parse(sql);
        assertThat(operation).isInstanceOf(ShowDatabasesOperation.class);
        final ShowDatabasesOperation showDatabasesOperation = (ShowDatabasesOperation) operation;
        assertThat(showDatabasesOperation.asSummaryString()).isEqualTo(expectedSummary);
    }
```


Moreover, we the catalog name provided in the constructor (`ShowDatabaseOperation`) is not necessarily the same as the one we obtain in `ShowDatabaseOperation::execute(Context ctx)`. That is why, we retrieve the catalog only if the `catalogName` is `null`:

```
        String cName =
                catalogName == null ? ctx.getCatalogManager().getCurrentCatalog() : catalogName;

```

dawidwys replied:
But, if the catalogName is null, doesn't it mean the user did not provide the name? What are the situation that `catalogName` can be null? Can it happen in any other situation than if there is no `FROM/IN catalog`?

To be honest the situation that the prepostion is not null, but the catalog name is empty looks like a bug to me. Is it an equivalent to `SHOW DATABASES FROM LIKE ...`? This doesn't look like a valid SQL statement.

jeyhunkarimov replied:
Sorry for the confusion. 
No, there is no case that `preposition` is `not null` and `catalog` is `null`. 
In fact, based on our discussion, I have already removed `preposition` field from `ShowDatabasesOperation`.

`catalog` is null (so the `preposition` that doesn't exist now) only when there is no `FROM/IN` clause. 

Do the changes in `ShowDatabasesOperation` reflect your comments or did I miss something?

dawidwys replied:
I am referring to the `SqlShowDatabasesConverter#convertSqlNode`:
```
        if (sqlShowDatabases.getPreposition() == null) {
...
        } else {
            String[] fullCatalogName = sqlShowDatabases.getCatalog();
            String catalogName =
                    fullCatalogName.length == 0
```

Is it possible that you have a preposition, but the catalog name is not present: `fullCatalogName.length == 0`. What is the equivalent SQL query that results in such a situation?

jeyhunkarimov replied:
ok, got it. Good point. No, if the preposition is not null, catalog name should be present. I updated `SqlShowDatabasesConverter `

dawidwys replied:
Great! Thank you!

dawidwys reviewed (commented):
Thank you @jeyhunkarimov for the work! The PR looks really good! I put some small comments. I think we can get it in shortly!

jeyhunkarimov commented:
@flinkbot run azure

jeyhunkarimov commented:
@flinkbot run azure

dawidwys reviewed (changes requested):
The changes look good. There is one last comment to be addressed and I'll merge the PR. Thank you @jeyhunkarimov

dawidwys reviewed (approved):
lgtm
